library/bignum.c

```diff
@@ -2280,6 +2280,21 @@ int mbedtls_mpi_gen_prime( mbedtls_mpi *X, size_t nbits, int dh_flag,
                 MBEDTLS_MPI_CHK( mbedtls_mpi_add_int( &Y, &Y, 6  ) );
             }
         }
+        /* check for RNG validity if flag is set (Note: this has a false
+        positive probability of %1). If more than 5 * nlen/2
+        (nlen/2 is the size of the random buffer in bits) generations fail
+        return an error indicating a possible issue with RNG.
+        */
+        if( ( flags & MBEDTLS_MPI_GEN_PRIME_CHECK_RNG ) ==
+            MBEDTLS_MPI_GEN_PRIME_CHECK_RNG )
+        {
+            rng_validity_counter++;
+            if ( rng_validity_counter > 5 * nbits )
+            {
+                ret = MBEDTLS_ERR_MPI_RNG_POSSIBLY_FAULTY;
+                goto cleanup;
+            }
+        }
```

Comment:
As of commit https://github.com/ARMmbed/mbedtls/pull/2036/commits/85eb58bf77e8cc4ef86892a076539e06c3037af8, this roughly implements the iteration count of FIPS 186-4 §B.3.3, but not exactly so. FIPS 186-4 does not count an iteration if p or q is rejected at the sqrt(2) check or at the |p-q| check, but it does count a rejection at the gcd check. The |p-q| check has a negligible impact (~2^{-100} probability of failing). The gcd check fails about 2^{-16} of the time with the typical choice e=65537 (which is the minimum permitted value), but can fail more often with a non-prime or small e. For example, for e=65541, the gcd check fails >½ the time since p-1 has a ½ chance of being divisible by 3 (we know that p isn't) and e is divisible by 3. This means that our procedure may have a significantly reduced chance of failure, compared to FIPS, but that's ok, functionally speaking.